### PR TITLE
[Wang Pei] iP

### PR DESCRIPTION
**Pull request by wangpalex**
### Duke v1.0 has already implemented basic funcionalities of a ToDo list:
- ~~nice~~ *ordinary* UI.
- track tasks, deadlines and events as you want.
- mark tasks done whenever you want.
- auto-save the file after changes.
- load task list from given file.
- find the wanted task with keywords.

### Command of Duke follows the format below:
1. todo `taskName` [-p `priority`]
2. deadline `deadlineName` /by `dateTime: yyyy/MM/dd HH:mm` [-p `priority`]
3. event `eventName` /at `startDateTime: yyyy/MM/dd HH:mm` `endDateTime: yyyy/MM/dd HH:mm` [-p `priority`]
4. list
5. done `index`
6. delete `index`
7. find `keyword`

### Features under working:
- [x] A-Gradle
- [x] A-Checkstyle
- [x] Level-10
- [x] A-Varargs

### If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/duke/Duke.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * Main logic of Duke,
+     * it starts the conversation loop and continuously reads input from
+     * user and execute corresponding functions until exit instructions are given.
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
Comment (chrus-chong) [inline src/main/java/duke/Duke.java]: Should this boolean variable be named in a more intuitive manner? Perhaps `hasExited` might be better?

File: src/main/java/duke/command/Command.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        this.isExit = false;
+    }
+
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
+
+    /**
+     * Returns a boolean indicating if the main loop should be broke.
+     * @return A boolean indicating if the main loop should be broke
+     */
+    public boolean exeResult() {
```
Comment (chrus-chong) [inline src/main/java/duke/command/Command.java]: Perhaps you may wish to name your `exeResult( )` method into something more intuitive? Was this meant to be shortform for `executeResult( )` ?

File: src/main/java/duke/command/CommandFind.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * Execute find action and then print result message.
+     * @param tasks The taskList to search keywork in.
+     * @param ui The Ui object to print messages.
+     * @param storage The Storage object that auto-saves after modification.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        TaskList matched = tasks.find(keyword);
```
Comment (chrus-chong) [inline src/main/java/duke/command/CommandFind.java]: Should this variable be renamed to `matchedTasks` for better identification?

File: src/main/java/duke/command/CommandAdd.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * The class that models an add command.
+ */
+public class CommandAdd extends Command {
+    Task taskToAdd;
+
+    /**
+     * Instantiates an object with given task to be added when
+     * called to execute.
+     * @param taskToAdd
+     */
```
Comment (chrus-chong) [inline src/main/java/duke/command/CommandAdd.java]: Consider adding a blank line between the method description and the `@param` statement to follow the Javadocs formatting in the [Java Coding Standard](https://nus-cs2103-ay1920s2.github.io/website/coding-standards/java/intermediate.html#variables).

**Review by chrus-chong — COMMENTED** [review]
Great job on the project! I cannot seem to find any recurring major coding standard violations. Keep up the good work!

File: src/main/java/duke/command/CommandAdd.java (hunk 4 above)
Comment (wangpalex) [reply src/main/java/duke/command/CommandAdd.java]: Hi, thanks for pointing out these defects so I can improve my code quality. I will make modifications according to suggestions! 😄

File: src/main/java/duke/Duke.java (hunk 1 above)
Comment (wangpalex) [reply src/main/java/duke/Duke.java]: Thanks for the suggestion! `hasExited` is indeed a better var name.

File: src/main/java/duke/command/Command.java (hunk 2 above)
Comment (wangpalex) [reply src/main/java/duke/command/Command.java]: Sure, I will use full description for this function name.

File: src/main/java/duke/command/CommandFind.java (hunk 3 above)
Comment (wangpalex) [reply src/main/java/duke/command/CommandFind.java]: Thanks for the suggestion!

File: src/main/java/duke/command/CommandAdd.java (hunk 4 above)
Comment (wangpalex) [reply src/main/java/duke/command/CommandAdd.java]: Noted, I will check out the coding standard again.

File: test/duke/util/ParserTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Test
+    void parseIndex() throws DukeException {
+        assertEquals(1, Parser.parseIndex(new String[] {"done", "1"}));
+    }
+
+    @Test
+    void parseDescription() throws DukeException {
+        assertEquals("description part 1 description part 2", Parser.parseDescription(new String[] {"taskType","description part 1", "description part 2"}));
```
Comment (bernardwan) [inline test/duke/util/ParserTest.java:51]: Should have a line break perhaps

File: src/main/java/duke/util/TaskList.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * Delete task at the specified index.
+     * @param idx The index of the task to be deleted.
+     * @return String of messages to be printed by Ui object after deleting.
+     * @throws DukeException When the index is invalid.
+     */
+    public String deleteTask(int idx) throws DukeException {
+        try {
+            StringBuilder msg = new StringBuilder();
+            Task curr = this.tasks.remove(idx-1);
```
Comment (bernardwan) [inline src/main/java/duke/util/TaskList.java]: Under advanced coding standard violations, it is said not to use 'this' for fields if it is not necessary. Can consider removing this here and other parts of code perhaps.

File: src/main/java/duke/command/CommandDone.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+     * Instantiates an object with given index to mark the task as done when
+     * called to execute.
+     * @param index The index where the task should be deleted.
+     */
+    public CommandDone(int index) {
+        this.index = index;
+    }
+
+    /**
+     * Execute the mark as done action and auto-saves to the file when called.
```
Comment (bernardwan) [inline src/main/java/duke/command/CommandDone.java:28]: For header comments, should leave a line in between the Description of the method and the description of the params.

**Review by bernardwan — COMMENTED** [review]
Naming seems very clear overall! Couldn't find much to comment on.

File: src/main/java/duke/Duke.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    private Storage storage;
+    private TaskList tasks;
+
+    public static void main(String[] args) {
+        Duke duke = new Duke(DEFAULT_PATH);
+        duke.run();
+    }
+
+    /**
+     * Main logic of Duke,
```
Comment (tsy24) [inline src/main/java/duke/Duke.java]: Consider starting the sentence with a verb like `Starts` to follow the [Java Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html)

File: src/main/java/duke/Duke.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        this.tasks = taskList;
+    }
+
+    /**
+     * Get response messages from Duke according input.
+     * @param input The input string to be parsed.
+     * @return      Execution message of the input or error messages.
+     * @throws DukeException when something goes wrong.
+     */
+
```
Comment (tsy24) [inline src/main/java/duke/Duke.java]: For Javadoc comments, the coding standard is to have an empty line between description and parameter section. Also, there should not be a blank line between the documentation block and the method.

File: src/main/java/duke/task/Event.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    public Event(String description, LocalDateTime startDateTime, LocalDateTime endDateTime) {
+        super(description);
+        this.startDateTime = startDateTime;
+        this.endDateTime = endDateTime;
+    }
+
+    /**
+     * Constructor.
+     * Instantiates a Deadline object with given description, isDone status and
```
Comment (tsy24) [inline src/main/java/duke/task/Event.java:59]: Could this be a typo? Did you mean Event object instead?

File: src/main/java/duke/ui/Ui.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    public String readLine() {
+        String input = scanner.nextLine();
+        return input;
+    }
+
+    /**
+     * Greets user.
+     * @param name User name to display.
+     */
+    public void greeting(String name) {
```
Comment (tsy24) [inline src/main/java/duke/ui/Ui.java]: Perhaps you can consider naming your method with a verb? Maybe something like `greetUser` or `sayGreeting`.

File: src/main/java/duke/util/Parser.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+    /**
+     * The method takes a user input and parse out all necessary parts needed
+     * to execute user command and finally returns a Command object that can be executed
+     * @param input A line of user input
+     * @return A Command object which can then be called to execute user command
+     * @throws DukeException When the input is empty String or the command is invalid
+     */
+    public static Command parseCommand(String input) throws DukeException {
+        String[] splited = input.strip().split(" +");
```
Comment (tsy24) [inline src/main/java/duke/util/Parser.java]: Plural form should be used on names representing a collection of objects. Perhaps `splitedStrings` would be a better name?

File: src/main/java/duke/util/Parser.java [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    /**
+     * Parse and return the task index of CommandDelete and CommandDone.
+     * @param strArr String array that contains splited input string.
+     * @return parsedIndex
+     * @throws DukeException when the index is missing from the input.
+     */
+    public static int parseIndex(String[] strArr) throws DukeException {
+        if(strArr.length < 2) {
```
Comment (tsy24) [inline src/main/java/duke/util/Parser.java]: According to the coding standard, Java reserved words should be followed by a white space, eg. `if (condition) {` and `for (i = 0; i < 10; i++) {` I noticed the same issue in several other places too.

File: test/duke/util/TaskListTest.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+import duke.exception.DukeException;
+import duke.task.Event;
+import duke.task.Task;
+import org.junit.jupiter.api.Test;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (tsy24) [inline test/duke/util/TaskListTest.java:12]: Consider listing imported classes explicitly to follow the coding standards.

File: src/main/java/duke/util/TaskList.java (hunk 6 above)
Comment (wangpalex) [reply src/main/java/duke/util/TaskList.java]: Hey @bernardwan, thanks for pointing that out, I will remove redandunt `this`.

File: src/main/java/duke/Duke.java (hunk 9 above)
Comment (wangpalex) [reply src/main/java/duke/Duke.java]: Thanks @tsy24, I would fix on that.

**Pull request closed without merging**